File: src/conversa/WebSocket.pas
// Eduardo - 31/07/2025
unit WebSocket;

interface

uses
  System.SysUtils,
  System.JSON,
  IdContext,
  Bird.Socket.Connection;

{$SCOPEDENUMS ON}

type
  TBirdSocketConnectionHack = class
  private
    FIdContext: TIdContext; // Mesmo layout de memória
  end;

  TSocketMessageType = (
    Erro,
    Login,
    NovaMensagem,
    AtualizacaoStatusMensagem,
    Digitando,
    GravandoAudio,
    ReacaoMensagem = 7,
    ConversaNova = 40, // Nova conversa
    ChamadaRecebida = 51, // Usuário inicia uma chamada
    ChamadaFinalizada = 52, // Usuário que criou, cancela a chamada antes mesmo de algum usuário entrar ou finaliza a chamada de modo forçado
    UsuarioRecusou = 53,
    UsuarioEntrou = 54,
    UsuarioSaiu = 55,
    VideoAtivado = 56,
    StatusUsuario = 60
  );

  TWebSocket = record
  private
    class procedure Enviar(const sUsuario: String; const oJSON: TJSONObject); static;
  public
    class procedure Iniciar(const iPort: Integer; const sJWTKey: String); static;
    class procedure Parar; static;
    class procedure NovaMensagem(const sUsuario, sTitulo, sMensagem: String); static;
    class procedure AtualizarStatusMensagem(const sUsuario: String; const iGrupo: Integer; const sMensagens: String); static;

    class function UsuarioConectado(const sUsuario: String): Boolean; static;
    class procedure ConversaNotificar(const Conversa, Remetente, Destinatario: Integer; const Msg: TSocketMessageType); static;
    class procedure ReacaoNotificar(const Conversa, Mensagem, Remetente, Destinatario: Integer; const Emoji, Acao: String); static;
    class procedure ChamadaNotificar(const Chamada, Remetente, Destinatario: Integer; const Msg: TSocketMessageType); static;
    class procedure StatusUsuarioNotificar(const UsuarioAlvo, UsuarioMudou: Integer; const Online: Boolean); static;
    class procedure NotificarContatosStatus(const UsuarioId: Integer; const Online: Boolean); static;
  end;

  TWSUser = class
    ID: String;
  end;

implementation

uses
  System.Classes,
  System.Generics.Collections,
  Data.DB,
  FireDAC.Comp.Client,
  Bird.Socket,
  IdSSLOpenSSL,
  JOSE.Context,
  JOSE.Core.JWT,
  JOSE.Consumer,
  Postgres;

var
  FWebSocket: TBirdSocket;
  FJWTKey: String;

function ContarConexoesUsuario(const sUsuario: String): Integer;
var
  Birds: TList<TBirdSocketConnection>;
  Bird: TBirdSocketConnection;
  Data: TObject;
  Context: TIdContext;
begin
  Result := 0;
  if sUsuario.IsEmpty then
    Exit;

  FWebSocket.Contexts.LockList;
  Birds := FWebSocket.Birds.LockList;
  try
    for Bird in Birds do
    begin
      Context := TBirdSocketConnectionHack(Bird).FIdContext;
      if not Assigned(Context) then
        Continue;
      Data := Context.Data;
      if Assigned(Data) and (Data is TWSUser) and (TWSUser(Data).ID = sUsuario) then
        Inc(Result);
    end;
  finally
    FWebSocket.Birds.UnLockList;
    FWebSocket.Contexts.UnlockList;
  end;
end;

procedure Execute(const ABird: TBirdSocketConnection);
var
  sMensagem: String;
  oJSON: TJSONObject;
  oErro: TJSONObject;
  User: TWSUser;
  LJWT: TJOSEContext;
  LBuilder: IJOSEConsumerBuilder;
  Context: TIdContext;
begin
  sMensagem := ABird.WaitMessage;
  if sMensagem.Trim.IsEmpty then
    Exit;

  oJSON := TJSONObject.ParseJSONValue(sMensagem) as TJSONObject;
  try
    if not Assigned(oJSON) or (oJSON.FindValue('tipo') = nil) then
    begin
      oErro := TJSONObject.Create;
      try
        oErro.AddPair('tipo', 9);
        if not Assigned(oJSON) then
          oErro.AddPair('message', 'Erro ao ler os dados do WebSocket: JSON inválido!')
        else
          oErro.AddPair('message', 'Erro ao ler os dados do WebSocket: Par "tipo" não encontrado!');
        ABird.Send(oErro.ToJSON);
      finally
        oErro.Free;
      end;
      Exit;
    end;

    case TSocketMessageType(oJSON.GetValue<Integer>('tipo')) of
      TSocketMessageType.Login:
      begin
        LJWT := TJOSEContext.Create(oJSON.GetValue<String>('token'), TJWTClaims);
        try
          LBuilder := TJOSEConsumerBuilder.NewConsumer;
          LBuilder.SetVerificationKey(FJWTKey);
          LBuilder.SetSkipVerificationKeyValidation;
          LBuilder.SetExpectedAudience(False, []);
          LBuilder.SetRequireExpirationTime;
          LBuilder.SetRequireIssuedAt;
          LBuilder.SetRequireSubject;

          try
            LBuilder.Build.ProcessContext(LJWT);
          except on E: Exception do
            begin
              oErro := TJSONObject.Create;
              try
                oErro.AddPair('tipo', Integer(TSocketMessageType.Erro));
                oErro.AddPair('message', E.Message);
                ABird.Send(oErro.ToJSON);
              finally
                oErro.Free;
              end;
              Exit;
            end;
          end;

          User := TWSUser.Create;
          User.ID := LJWT.GetClaims.Subject;

          FWebSocket.Contexts.LockList;
          try
            // Acessa o campo FIdContext protegido da outra classe mapeando a mesma posição na classe hack
            Context := TBirdSocketConnectionHack(ABird).FIdContext;
            if Assigned(Context) and not Assigned(Context.Data) then
              Context.Data := User;
          finally
            FWebSocket.Contexts.UnlockList;
          end;

          // Notificar contatos que o usuário ficou online (só na primeira conexão)
          if ContarConexoesUsuario(User.ID) <= 1 then
            TWebSocket.NotificarContatosStatus(User.ID.ToInteger, True);
        finally
          LJWT.Free;
        end;
      end;
    end;
  finally
    oJSON.Free;
  end;
end;

procedure Disconnect(const ABird: TBirdSocketConnection);
var
  Context: TIdContext;
  Data: TObject;
  UserId: Integer;
  sUserId: String;
begin
  Context := TBirdSocketConnectionHack(ABird).FIdContext;
  if not Assigned(Context) then
    Exit;

  Data := Context.Data;
  if not Assigned(Data) or not (Data is TWSUser) then
    Exit;

  sUserId := TWSUser(Data).ID;
  UserId := StrToIntDef(sUserId, 0);

  // Só notifica offline se esta é a última conexão do usuário
  // (ContarConexoesUsuario conta incluindo esta que ainda está na lista)
  if (UserId > 0) and (ContarConexoesUsuario(sUserId) <= 1) then
    TWebSocket.NotificarContatosStatus(UserId, False);
end;

class procedure TWebSocket.Iniciar(const iPort: Integer; const sJWTKey: String);
begin
  FJWTKey := sJWTKey;
  FWebSocket := TBirdSocket.Create(iPort);
  FWebSocket.Active := True;
  FWebSocket.AddEventListener(TEventType.EXECUTE, Execute);
  FWebSocket.AddEventListener(TEventType.DISCONNECT, Disconnect);
end;

class procedure TWebSocket.Enviar(const sUsuario: String; const oJSON: TJSONObject);
var
  Birds : TList<TBirdSocketConnection>;
  Bird: TBirdSocketConnection;
  Data: TObject;
  Context: TIdContext;
begin
  if sUsuario.IsEmpty or not Assigned(oJSON) then
    Exit;

  FWebSocket.Contexts.LockList;
  Birds := FWebSocket.Birds.LockList;
  try
    for Bird in Birds do
    begin
      // Acessa o campo FIdContext protegido da outra classe mapeando a mesma posição na classe hack
      Context := TBirdSocketConnectionHack(Bird).FIdContext;
      if not Assigned(Context) then
        Continue;

      Data := Context.Data;

      if not Assigned(Data) or not (Data is TWSUser) or (TWSUser(Data).ID <> sUsuario) then
        Continue;

      Bird.Send(oJSON.ToJSON);
    end;
  finally
    FWebSocket.Birds.UnLockList;
    FWebSocket.Contexts.UnlockList;
  end;
end;

class function TWebSocket.UsuarioConectado(const sUsuario: String): Boolean;
var
  Birds: TList<TBirdSocketConnection>;
  Bird: TBirdSocketConnection;
  Data: TObject;
  Context: TIdContext;
begin
  Result := False;
  if sUsuario.IsEmpty then
    Exit;

  FWebSocket.Contexts.LockList;
  Birds := FWebSocket.Birds.LockList;
  try
    for Bird in Birds do
    begin
      Context := TBirdSocketConnectionHack(Bird).FIdContext;
      if not Assigned(Context) then
        Continue;

      Data := Context.Data;

      if Assigned(Data) and (Data is TWSUser) and (TWSUser(Data).ID = sUsuario) then
        Exit(True);
    end;
  finally
    FWebSocket.Birds.UnLockList;
    FWebSocket.Contexts.UnlockList;
  end;
end;

class procedure TWebSocket.NovaMensagem(const sUsuario, sTitulo, sMensagem: String);
var
  oJSON: TJSONObject;
begin
  oJSON := TJSONObject.Create;
  try
    oJSON.AddPair('tipo', Integer(TSocketMessageType.NovaMensagem));
    oJSON.AddPair('titulo', sTitulo);
    oJSON.AddPair('mensagem', sMensagem);
    TWebSocket.Enviar(sUsuario, oJSON);
  finally
    oJSON.Free;
  end;
end;

class procedure TWebSocket.AtualizarStatusMensagem(const sUsuario: String; const iGrupo: Integer; const sMensagens: String);
var
  oJSON: TJSONObject;
begin
  oJSON := TJSONObject.Create;
  try
    oJSON.AddPair('tipo', Integer(TSocketMessageType.AtualizacaoStatusMensagem));
    oJSON.AddPair('grupo', iGrupo);
    oJSON.AddPair('mensagens', sMensagens);
    TWebSocket.Enviar(sUsuario, oJSON);
  finally
    oJSON.Free;
  end;
end;

class procedure TWebSocket.ConversaNotificar(const Conversa, Remetente, Destinatario: Integer; const Msg: TSocketMessageType);
var
  jo: TJSONObject;
begin
  jo := TJSONObject.Create;
  try
    jo.AddPair('tipo', Integer(Msg));
    jo.AddPair('conversa_id', Conversa);
    jo.AddPair('usuario_id', Remetente);
    TWebSocket.Enviar(Destinatario.ToString, jo);
  finally
    jo.Free;
  end;
end;

class procedure TWebSocket.ReacaoNotificar(const Conversa, Mensagem, Remetente, Destinatario: Integer; const Emoji, Acao: String);
var
  jo: TJSONObject;
begin
  jo := TJSONObject.Create;
  try
    jo.AddPair('tipo', Integer(TSocketMessageType.ReacaoMensagem));
    jo.AddPair('conversa_id', Conversa);
    jo.AddPair('mensagem_id', Mensagem);
    jo.AddPair('usuario_id', Remetente);
    jo.AddPair('emoji', Emoji);
    jo.AddPair('acao', Acao);
    TWebSocket.Enviar(Destinatario.ToString, jo);
  finally
    jo.Free;
  end;
end;

class procedure TWebSocket.ChamadaNotificar(const Chamada, Remetente, Destinatario: Integer; const Msg: TSocketMessageType);
var
  jo: TJSONObject;
begin
  jo := TJSONObject.Create;
  try
    jo.AddPair('tipo', Integer(Msg));
    jo.AddPair('chamada_id', Chamada);
    jo.AddPair('usuario_id', Remetente);
    TWebSocket.Enviar(Destinatario.ToString, jo);
  finally
    jo.Free;
  end;
end;

class procedure TWebSocket.StatusUsuarioNotificar(const UsuarioAlvo, UsuarioMudou: Integer; const Online: Boolean);
var
  jo: TJSONObject;
begin
  jo := TJSONObject.Create;
  try
    jo.AddPair('tipo', Integer(TSocketMessageType.StatusUsuario));
    jo.AddPair('usuario_id', UsuarioMudou);
    jo.AddPair('online', TJSONBool.Create(Online));
    TWebSocket.Enviar(UsuarioAlvo.ToString, jo);
  finally
    jo.Free;
  end;
end;

class procedure TWebSocket.NotificarContatosStatus(const UsuarioId: Integer; const Online: Boolean);
var
  Pool: IConnection;
  Qry: TFDQuery;
begin
  Pool := TPool.Instance;
  Qry := TFDQuery.Create(nil);
  try
    Qry.Connection := Pool.Connection;
    // Busca todos os usuários que compartilham conversas diretas (tipo=1) com o usuário
    Qry.Open(
      'select distinct cu2.usuario_id '+
      '  from conversa_usuario cu1 '+
      ' inner join conversa c on c.id = cu1.conversa_id and c.tipo = 1 '+
      ' inner join conversa_usuario cu2 on cu2.conversa_id = cu1.conversa_id and cu2.usuario_id <> cu1.usuario_id '+
      ' where cu1.usuario_id = '+ UsuarioId.ToString
    );

    if Qry.IsEmpty then
      Exit;

    Qry.FetchAll;
    Qry.First;
    while not Qry.Eof do
    try
      TWebSocket.StatusUsuarioNotificar(Qry.FieldByName('usuario_id').AsInteger, UsuarioId, Online);
    finally
      Qry.Next;
    end;
  finally
    FreeAndNil(Qry);
  end;
end;

class procedure TWebSocket.Parar;
begin
  FWebSocket.Free;
end;

end.

File: src/websocket.ts
import websocket from '@fastify/websocket'
import type { FastifyInstance } from 'fastify'
import type { WebSocket } from 'ws'
import { comUsuario } from './banco.ts'

export const TipoMensagemSocket = {
  Erro: 0,
  Login: 1,
  NovaMensagem: 2,
  AtualizacaoStatusMensagem: 3,
  Digitando: 4,
  GravandoAudio: 5,
  ReacaoMensagem: 7,
  ConversaNova: 40,
  ChamadaRecebida: 51,
  ChamadaFinalizada: 52,
  UsuarioRecusou: 53,
  UsuarioEntrou: 54,
  UsuarioSaiu: 55,
  VideoAtivado: 56,
  StatusUsuario: 60,
} as const
export type TipoMensagemSocket = (typeof TipoMensagemSocket)[keyof typeof TipoMensagemSocket]

// Conexoes abertas por usuario. Um usuario pode ter varias abas e aparelhos.
const conexoes = new Map<number, Set<WebSocket>>()

type VerificarToken = (token: string) => { sub: string }

// WebSocket na mesma porta da API, em /ws/. A autenticacao nao usa o
// cabecalho: o cliente manda { tipo: 1, token } como primeira mensagem.
export async function registrarWebSocket(app: FastifyInstance, verificarToken: VerificarToken) {
  await app.register(websocket)

  app.get('/ws/', { websocket: true }, (socket) => {
    let usuarioId = 0

    socket.on('message', (bruto) => {
      let mensagem: { tipo?: unknown; token?: unknown } | undefined
      try {
        mensagem = JSON.parse(bruto.toString())
      } catch {
        mensagem = undefined
      }

      if (!mensagem || typeof mensagem !== 'object' || mensagem.tipo === undefined) {
        socket.send(JSON.stringify({
          tipo: 9,
          message: mensagem
            ? 'Erro ao ler os dados do WebSocket: Par "tipo" não encontrado!'
            : 'Erro ao ler os dados do WebSocket: JSON inválido!',
        }))
        return
      }

      if (mensagem.tipo !== TipoMensagemSocket.Login || usuarioId) {
        return
      }

      try {
        usuarioId = Number(verificarToken(String(mensagem.token ?? '')).sub)
      } catch (erro) {
        socket.send(JSON.stringify({ tipo: TipoMensagemSocket.Erro, message: erro instanceof Error ? erro.message : String(erro) }))
        return
      }

      let doUsuario = conexoes.get(usuarioId)
      if (!doUsuario) {
        conexoes.set(usuarioId, doUsuario = new Set())
      }
      doUsuario.add(socket)
      // Avisa os contatos so na primeira conexao do usuario.
      if (doUsuario.size === 1) {
        notificarContatosStatus(usuarioId, true)
      }
    })

    socket.on('close', () => {
      if (!usuarioId) {
        return
      }
      const doUsuario = conexoes.get(usuarioId)
      doUsuario?.delete(socket)
      // Offline so quando fecha a ultima conexao.
      if (!doUsuario?.size) {
        conexoes.delete(usuarioId)
        notificarContatosStatus(usuarioId, false)
      }
    })
  })
}

function enviar(usuarioId: number, dados: object) {
  const destinos = conexoes.get(usuarioId)
  if (!destinos) {
    return
  }
  const texto = JSON.stringify(dados)
  for (const socket of destinos) {
    socket.send(texto)
  }
}

export function usuarioConectado(usuarioId: number) {
  return (conexoes.get(usuarioId)?.size ?? 0) > 0
}

export function notificarNovaMensagem(usuarioId: number, titulo: string, mensagem: string) {
  enviar(usuarioId, { tipo: TipoMensagemSocket.NovaMensagem, titulo, mensagem })
}

export function notificarStatusMensagem(usuarioId: number, conversa: number, mensagens: string) {
  enviar(usuarioId, { tipo: TipoMensagemSocket.AtualizacaoStatusMensagem, grupo: conversa, mensagens })
}

export function notificarConversa(conversa: number, remetente: number, destinatario: number, tipo: TipoMensagemSocket) {
  enviar(destinatario, { tipo, conversa_id: conversa, usuario_id: remetente })
}

export function notificarReacao(conversa: number, mensagem: number, remetente: number, destinatario: number, emoji: string, acao: string) {
  enviar(destinatario, { tipo: TipoMensagemSocket.ReacaoMensagem, conversa_id: conversa, mensagem_id: mensagem, usuario_id: remetente, emoji, acao })
}

export function notificarChamada(chamada: number, remetente: number, destinatario: number, tipo: TipoMensagemSocket) {
  enviar(destinatario, { tipo, chamada_id: chamada, usuario_id: remetente })
}

// Avisa quem tem conversa direta com o usuario que ele entrou ou saiu.
function notificarContatosStatus(usuarioId: number, online: boolean) {
  comUsuario(0, (sql) => sql<{ usuario_id: number }[]>`
    select distinct cu2.usuario_id
      from conversa_usuario cu1
     inner join conversa c on c.id = cu1.conversa_id and c.tipo = 1
     inner join conversa_usuario cu2 on cu2.conversa_id = cu1.conversa_id and cu2.usuario_id <> cu1.usuario_id
     where cu1.usuario_id = ${usuarioId}`)
    .then((contatos) => {
      for (const { usuario_id } of contatos) {
        enviar(usuario_id, { tipo: TipoMensagemSocket.StatusUsuario, usuario_id: usuarioId, online })
      }
    })
    .catch((erro) => console.error('[WebSocket] Falha ao notificar status do usuário', usuarioId, erro))
}
